amethyst_core/src/entity_uuid.rs

```diff
@@ -0,0 +1,157 @@
+//! This module provides the EntityUuid component, and the EntityUuidMap resource.
+//!
+//! Together, these can be used to create entity identifiers that can persist across game sessions, and across machines.
+//!
+//! These properties make them useful for serializing a reference to an entity to disk, such as saving the game, or for
+//! identification in a networked multiplayer situation.
+
+use fnv::FnvHashMap;
+use hibitset::BitSetLike;
+use shrev::ReaderId;
+use specs::{
+    storage::ComponentEvent, BitSet, Component, Entities, Entity, FlaggedStorage, ReadStorage,
+    Resources, System, SystemData, Write, WriteStorage,
+};
+use uuid::Uuid;
+
+/// An identifier for an entity which can persist across game sessions, and across machines.
+///
+/// These properties make it useful for serializing a reference to an entity to disk, such as saving the game, or for
+/// identification in a networked multiplayer situation.
+///
+/// Once the Uuid is initialized in this structure it should not be altered.
+#[derive(Debug, Eq, PartialEq)]
+pub struct EntityUuid {
+    uuid: Uuid,
+}
+
+impl Component for EntityUuid {
+    type Storage = FlaggedStorage<Self>;
+}
+
+impl EntityUuid {
+    /// Create a new instance with a new randomly generated Uuid.
+    fn new() -> Self {
+        Self {
+            uuid: Uuid::new_v4(),
+        }
+    }
+
+    /// Create a new instance with a pre-defined Uuid, typically the Uuid would be deserialized from the disk or network.
+    fn new_from_uuid(uuid: Uuid) -> Self {
+        Self { uuid }
+    }
+
+    /// Retrieve the Uuid structure contained in this component.
+    pub fn uuid(&self) -> &Uuid {
+        &self.uuid
+    }
+}
+
+/// An ECS resource which allows you to create new EntityUuid components, and later retrieve the entities with the contained UUID.
+#[derive(Debug, Default)]
+pub struct EntityUuidMap {
+    map: FnvHashMap<Uuid, Entity>,
+}
+
+impl EntityUuidMap {
+    /// Create an empty instance of this resource.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Create a new EntityUuid component with a new randomly generated Uuid.
+    pub fn new_uuid(&mut self, e: Entity) -> EntityUuid {
+        let r = EntityUuid::new();
+        self.map.insert(r.uuid, e);
+        r
+    }
+
+    /// Create a new EntityUuid component with a pre-defined Uuid, typically the Uuid would be deserialized from the disk or network.
+    pub fn with_uuid(&mut self, uuid: Uuid, e: Entity) -> EntityUuid {
+        self.map.insert(uuid, e);
```

Comment:
Instead of this, I think you should have the System listen to inserted component events. This way it would prevent people from manually creating the component and inserting it on the entity without going through the map.

Reply:
The functions to manually create the component are not public, meaning that behavior is already prevented. You can only create the component through the map.